File: csharp/Dispatcher/Wrappers/Statusline.cs
using System.Text.Json;
using System.Text.Json.Nodes;

namespace LlmDispatcher.Wrappers;

/// <summary>
/// <c>llm-dispatcher statusline</c>: Claude Code runs this and pipes session JSON on stdin.
/// Prints the last routing decision so the user can see which model actually served the turn.
/// </summary>
public static class Statusline
{
    public static async Task RunAsync()
    {
        JsonObject input = new();
        try
        {
            if (Console.IsInputRedirected)
            {
                var raw = await Console.In.ReadToEndAsync();
                if (raw.Trim().Length > 0) input = JsonNode.Parse(raw) as JsonObject ?? new JsonObject();
            }
        }
        catch (JsonException) { /* no usable stdin */ }

        var cwd = (input["workspace"] as JsonObject)?["current_dir"]?.GetValue<string>() ?? input["cwd"]?.GetValue<string>() ?? Directory.GetCurrentDirectory();
        var pct = (input["context_window"] as JsonObject)?["used_percentage"] is JsonValue p && p.TryGetValue<double>(out var d) ? d : (double?)null;
        var requested = (input["model"] as JsonObject)?["display_name"]?.GetValue<string>() ?? "";
        var parts = new List<string>();

        JsonObject? status = null;
        try { status = JsonNode.Parse(File.ReadAllText(TurnRouter.ClaudeStatusFile)) as JsonObject; }
        catch (Exception ex) when (ex is IOException or JsonException) { /* no decision yet */ }

        if (requested.Length > 0 && !requested.Contains("dispatcher", StringComparison.OrdinalIgnoreCase))
            parts.Add($"⏸ manual {requested}");
        else if (status is not null)
        {
            var model = (status["model"]?.GetValue<string>() ?? "").Replace("anthropic/", "");
            var effort = status["effort"]?.GetValue<string>();
            var conf = (status["confidence"] as JsonObject)?["difficulty"]?.GetValue<double>();
            var source = status["source"]?.GetValue<string>();
            var src = source == "jev" ? $"jev{(conf is { } c ? $" p={c:F2}" : "")}" : source ?? "";
            parts.Add($"⚡ {model}{(effort is null ? "" : $" @{effort}")} ({status["task"] ?? "?"}, {src})");
        }
        else parts.Add("⚡ dispatcher: waiting for first turn");

        parts.Add(Path.GetFileName(cwd.TrimEnd(Path.DirectorySeparatorChar)));
        if (pct is { } used) parts.Add($"{Math.Round(used)}% context");
        Console.Out.Write(string.Join(" · ", parts));
    }
}

File: ts/src/wrappers/statusline.ts
import { readFileSync } from "node:fs";
import { basename } from "node:path";
import { CLAUDE_STATUS_FILE } from "./session.js";

/**
 * `llm-dispatcher statusline`: Claude Code runs this and pipes session JSON on stdin.
 * Prints the last routing decision so the user can see which model actually served the turn.
 */
export async function statusline(): Promise<void> {
  let input: Record<string, unknown> = {};
  try {
    const raw = readFileSync(0, "utf8");
    if (raw.trim()) input = JSON.parse(raw);
  } catch {
    /* no stdin */
  }
  const cwd = (input.workspace as Record<string, string> | undefined)?.current_dir ?? (input.cwd as string | undefined) ?? process.cwd();
  const pct = (input.context_window as Record<string, number> | undefined)?.used_percentage;
  const requested = (input.model as Record<string, string> | undefined)?.display_name ?? "";
  const parts: string[] = [];

  let status: Record<string, unknown> | undefined;
  try {
    status = JSON.parse(readFileSync(CLAUDE_STATUS_FILE, "utf8"));
  } catch {
    /* no decision yet */
  }
  if (requested && requested !== "Dispatcher" && !/dispatcher/i.test(requested)) {
    parts.push(`⏸ manual ${requested}`);
  } else if (status) {
    const model = String(status.model ?? "").replace(/^anthropic\//, "");
    const effort = status.effort ? ` @${status.effort}` : "";
    const conf = (status.confidence as Record<string, number> | undefined)?.difficulty;
    const src = status.source === "jev" ? `jev${conf !== undefined ? ` p=${conf.toFixed(2)}` : ""}` : String(status.source ?? "");
    parts.push(`⚡ ${model}${effort} (${status.task ?? "?"}, ${src})`);
  } else {
    parts.push("⚡ dispatcher: waiting for first turn");
  }
  parts.push(basename(cwd));
  if (typeof pct === "number") parts.push(`${Math.round(pct)}% context`);
  process.stdout.write(parts.join(" · "));
}
